crates/harp/src/fixtures/mod.rs

```diff
@@ -27,13 +21,10 @@ use stdext::cargs;
 use crate::library::RLibraries;
 use crate::R_MAIN_THREAD_ID;
 
-// Escape hatch for unit tests. We need this because the default
-// implementation of `r_task()` needs a fully formed `RMain` to send the
-// task to, which we don't have in unit tests. Consequently tasks run
-// immediately in the current thread in unit tests. Since each test has its
-// own thread, they are synchronised via the `R_RUNTIME_LOCK` mutex.
-pub static mut R_TASK_BYPASS: bool = false;
-static mut R_RUNTIME_LOCK: Mutex<()> = Mutex::new(());
+// This needs to be a reentrant mutex because many of our tests are wrapped in
+// `r_test()` which takes the R lock. Without a reentrant mutex, we'd get
+// deadlocked when we cause some other background thread to use an `r_task()`.
```

Comment:
Since we're no longer wrapping in `r_test()`, I removed the reentrant lock. We only need reentrancy in production builds to avoid deadlocking if there is an unexpected nested `r_task()`. But that should not happen: either we are on the main thread, or we are not. A nested `r_task()` indicates code that is confused about where it should run.

I think we may have initially envisioned having tools that could be called both from the R thread and background threads, but that was misguided. The `r_task()` should be explicit and only used in background thread code.

Reply:
Well it's now a reentrant lock again for https://github.com/posit-dev/ark/pull/541#discussion_r1782580736